rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java

```diff
@@ -725,24 +725,30 @@ protected LRAStatus toLRAStatus(int atomicActionStatus) {
     }
 
     public LRAParticipantRecord enlistParticipant(URI coordinatorUrl, String participantUrl, String recoveryUrlBase,
-                                                  long timeLimit, String compensatorData) throws UnsupportedEncodingException {
-        LRAParticipantRecord participant = findLRAParticipant(participantUrl, false);
-
-        if (participant != null) {
-            participant.setCompensatorData(compensatorData);
-            return participant; // must have already been enlisted
-        }
-
-        participant = doEnlistParticipant(coordinatorUrl, participantUrl, recoveryUrlBase,
-                timeLimit, compensatorData);
-
-        if (participant != null) {
-            // need to remember that there is a new participant
-            deactivate(); // if it fails the superclass will have logged a warning
-            savedIntentionList = true; // need this clean up if the LRA times out
+            long timeLimit, String compensatorData) throws UnsupportedEncodingException {
+        ReentrantLock lock = tryLockTransaction();
+        if (lock != null) {
+            try {
+                LRAParticipantRecord participant = findLRAParticipant(participantUrl, false);
+                if (participant != null) {
+                    participant.setCompensatorData(compensatorData);
+                    return participant; // must have already been enlisted
+                }
+                participant = doEnlistParticipant(coordinatorUrl, participantUrl, recoveryUrlBase, timeLimit,
+                        compensatorData);
+                if (participant != null) {
+                    // need to remember that there is a new participant
+                    deactivate(); // if it fails the superclass will have logged a warning
+                    savedIntentionList = true; // need this clean up if the LRA times out
+                }
+                return participant;
+            }
+            finally {
+                lock.unlock();
+            }
         }
-
-        return participant;
+        LRALogger.i18nLogger.error_enlistment();
```

Comment:
> it should also be reported if the internal call to doEnlistParticipant fails, if it isn't already
It is already checked, in fact in that case the participant is null and the caller returns an error: see https://github.com/jbosstm/narayana/blob/main/rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/service/LRAService.java#L393-L395

Reply:
Note also that we used ERROR for reporting the problem back to the client but WARN for logging inside the coordinator since it is an error for the client but not for the coordinator.

Reply:
Sorry Mike, I misinterpreted your sentence 'it is an error'. Now I clearly see your reasoning. I will set the level as WARN. Thanks